### ISSUE TITLE
Lei Mao's Log Book  Layer Normalization Explained

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4889]
Lei Mao's Log Book – Layer Normalization Explained<br>
<br>
https://ift.tt/2ym55KK<br>
<br>
Twitter<br>
<br>
<div>
<div> <h3>Introduction</h3> <p>Recently I came across with layer normalization in the Transformer model for machine translation and I found that a special normalization layer called “layer normalization” was used throughout the model, so I decided to check how it works and compare it with the batch normalization we normally used in computer vision models.</p> <h3>Mathematical Definition</h3> <p>Given inputs $x$ over a minibatch of size $m$, $B = \{x_1, x_2, …, x_m\}$, each sample $x_i$ contains $K$ elements, i.e. the length of flatten $x_i$ is $K$, by applying transformation of your inputs using some learned parameters $\gamma$ and $\beta$, the outputs could be expressed as $B’ = \{y_1, y_2, …, y_m\}$, where $y_i = {\text{LN}}_{\gamma, \beta} (x_i)$.</p> <p><br></p> <p>More concretely, we first calculate the mean and the variance of of each sample from the minibatch. For sample $x_i$ whose flatten format is $\{x_{i,1}, x_{i,2}, …, x_{i,K}\}$, we have its mean $\mu_i$ and variance $\sigma_i^2$.</p> <p>Then we normalize each sample such that the elements in the sample have zero mean and unit variance. $\epsilon$ is for numerical stability in case the denominator becomes zero by chance.</p> <p>Finally, there is a scaling and shifting step. $\gamma$ and $\beta$ are learnable parameters.</p> <p>We can see from the math above that layer normalization has nothing to do with other samples in the batch.</p> <h3>Layer Normalization vs Batch Normalization</h3> <p>I had a simple blog post on <a href="https://leimao.github.io/blog/Batch-Normalization/">batch normalization</a> previously. Like that simple blog post, I am not going to talk about the advantage of layer normalization over batch normalization or how to choose normalization techniques in this blog post.</p> <p><br></p> <p>One of the major differences in practice is that layer normalization does not have to use “running mean” and “running variance” since batch does not play a role in the computation.</p> <p><br></p> <p>If layer normalization is working on the outputs from a convolution layer, the math has to be modified slightly since it does not make sense to group all the elements from three distinct channels together and compute the mean and variance. Each channel is considered as an “independent” sample and all the normalization was done for that specific channel only within the sample.</p> <p><br></p> <p>Assume the input tensor has shape $[m, H, W, C]$, for each channel $c \in \{1,2, \cdots, C\}$</p> <p>Specifically for each channel, we have learnable parameters $\gamma_c$ and $\beta_c$, such that</p> <h3>Layer Normalization vs Instance Normalization ?</h3> <p>Instance normalization, however, only exists for 3D or higher dimensional tensor inputs, since it requires the tensor to have batch and each sample in the batch needs to have layers (channels). If the samples in batch only have 1 channel (a dummy channel), instance normalization on the batch is exactly the same to layer normalization on the batch with this single dummy channel removed. Batch normalization and layer normalization works for 2D tensors which only consists of batch dimension without layers. Surprisingly (or not?), instance normalization for 3D or 4D tensor is exactly the same to layer normalization for convolution outputs as I mentioned above, because each sample in the batch is an instance, we are layer normalizing samples which happen to have multiple channels, and we ignore batches during normalization. So we could say about instance normalization in this way, instance normalization is a natural extension of layer normalization to convolutions, or it is just a new name for an old concept.</p> <h3>References</h3> <hr>
<footer><div> <img alt="Lei Mao bio photo" src="https://leimao.github.io/images//author_images/Lei-Bio-Medium.jpg"><h3>Lei Mao</h3> <p>Machine Learning, Artificial Intelligence. On the Move.</p> <a href="http://twitter.com/matchaleimao" target="_blank"> Twitter</a> <a href="http://facebook.com/dukeleimao" target="_blank"><i></i> Facebook</a> <a href="http://linkedin.com/in/lei-mao" target="_blank"><i></i> LinkedIn</a> <a href="http://github.com/leimao" target="_blank"><i></i> GitHub</a> <a href="http://scholar.google.com/citations?user=R2VUf7YAAAAJ" target="_blank"><i></i>  G. Scholar</a> <a href="mailto:dukeleimao@gmail.com" target="_blank"><i></i> E-Mail</a> </div> <p><strong>Layer Normalization Explained</strong> was published on <time datetime="2019-05-31T00:00:00-05:00">May 31, 2019</time> and last modified on <time datetime="2019-05-31">May 31, 2019</time> by <a title="About Lei Mao" href="https://leimao.github.io">Lei Mao</a>.</p> </footer>
</div> </div><br>
<br>
<br>
<br>
<br>
<br>
via Lei Mao's Log Book https://ift.tt/2MtcO1I<br>
<br>
July 31, 2019 at 10:56AM
